File: 1001-1100/1087-Brace Expansion/1087-Brace Expansion.py
class Solution:
    def expand(self, S: str) -> List[str]:
        result = []
        def dfs(start, path):
            if start == len(S):
                result.append(path)
                return
            
            if S[start] == '{':
                index = S.find('}', start + 1)
                for c in sorted(S[start + 1:index].split(',')):
                    dfs(index + 1, path + c)
            else:
                dfs(start + 1, path + S[start])
        
        dfs(0, '')
        return result

File: 1001-1100/1087-Brace Expansion/1087-Brace Expansion.cpp
class Solution {
public:
    vector<string> expand(string S) {
        vector<string> result;
        string path;
        dfs(S, 0, path, result);
        return result;
    }
    
private:
    void dfs(string& S, int start, string path, vector<string>& result) {
        if (start == S.size()) {
            result.push_back(path);
            return;
        }
        
        if (S[start] == '{') {
            int index = -1;
            vector<char> options;
            for (int i = start + 1; i < S.size(); ++i) {
                if (S[i] == '}') {
                    index = i;
                    break;
                }
                else if (S[i] != ',') {
                    options.push_back(S[i]);
                }
            }

            sort(options.begin(), options.end());
            for (char c : options) {
                dfs(S, index + 1, path + c, result);
            }
        }
        else {
            dfs(S, start + 1, path + S[start], result);
        }
    }
};
